documentation of ie performance

david-gang opened the pull request:
This is a PR for #8616 

Please feel free to edit the document and place it where you find it appropriate. 
I did not include the css fixes included in #8329 because i did not have time to test these.

DevVersion commented on docs/app/performance/internet-explorer.md:
```diff
@@ -0,0 +1,75 @@
+	}).
+    run(function($animate){
+    	$animate.enabled(!isIE);
+    });
```
This is very nitpicky, the indention here looks really weird.


```js
var isIE = isBrowserIE();

angular
  .module('myApp', ['ngMaterial'])
  .config(function($mdInkRippleProvider) {
    if (isIE) {
      $mdInkRippleProvider.disableInkRipple();
    } 
  })
  .run(function($animate) {
    $animate.enabled(!isIE);
});
```

DevVersion commented on docs/app/performance/internet-explorer.md:
```diff
@@ -0,0 +1,75 @@
+angular.module('myApp', ['ngMaterial'])
+	.config(function($mdThemingProvider) {
+	  $mdThemingProvider.disableTheming();
+	});
```
Indention is here wrong as well.

DevVersion commented on docs/app/performance/internet-explorer.md:
````diff
@@ -0,0 +1,75 @@
+* Open the chrome developer tools
+* Execute the following javascript - This copies the theming css to the clipboard
+```javascript
+copy([].slice.call(document.styleSheets).map(e=>e.ownerNode).filter(e => e.attributes && e.attributes[0].nodeName==='md-theme-style').map(e => e.textContent).join("\n"))
````
Can we put each _stream_ statement in a new line, like
```js
copy([].slice.call(document.styleSheets)
  .map(e => e.ownerNode)
  .filter(e => e.hasAttribute('md-theme-style'))
  .map(e => e.textContent)
  .join('\n')
```

ThomasBurleson commented on docs/app/performance/internet-explorer.md:
```diff
@@ -0,0 +1,80 @@
+One of the greatest issues with Internet Explorer, is the slow layout calculation. The best solution is to keep the DOM small.
+
+This can be achieved by using [Virtual Repeat](https://material.angularjs.org/latest/demo/virtualRepeat)
+The solution works unfortunately just when the height of a single list item is fixed.
```
I think we should elaborate with more details here. And an example link or source snippet.

david-gang replied:
I agree. I will prepare an example where we basically take a long list with ng-repeat and convert it to md-virtual-repeat. I will explain that the virtual repeat essentially updates the viewport so that just the visible part of the list exists in the dom. Do you think about other aspects we may elaborate here?

ThomasBurleson replied:
By reusing DOM elements during scrolling (instead of creating new DOM elements), mdVR supports real-time scrolling through millions of records (assuming all in-memory).

ThomasBurleson commented on docs/app/performance/internet-explorer.md:
```diff
@@ -0,0 +1,80 @@
+
+##Animations
+
+Animations can be turned off globally by using the $animate service
```
This seems like a radical alternative. 

- Is it all animations?
- Only Ink Ripples ?
- Any other animations ?

david-gang replied:
I agree it is very radical. We need a fine grained approach here. Unfortunately i just know how to switch off ink ripples with the $mdInkRippleProvider. I am not aware of an option otherwise to selectively switch off the animations beside the big $animation hammer

topherfangio replied:
Also note that many components currently *rely* on animations to properly function, so turning them off globally may actually cause some components to fail (I know the speed dial is one of them, but there are others).

graphefruit replied:
I just disabled all `transitions` on the fab-speed-dail and its working. Just no animation, made a short gif:
![speeddail](https://cloud.githubusercontent.com/assets/6260244/15851954/890f5ac2-2c9f-11e6-9fda-4792674bb7b6.gif)

david-gang replied:
Hi @graphefruit ,
How do you disable the transitions?
Through javascript or css?

Thanks,
David

graphefruit replied:
I'm just using CSS-Disabling.
Like in #8329 here a sample:
```
md-checkbox ._md-icon {
    transition: none !important;
    transition-duration: 0ms !important;
    transition-delay: 0ms !important;
}
```
Just setting this to `none!important`. The duration and delay I set to 0ms (don't know if its realy needed)

Before this I tried to realy "destroy" the animation, but like topherfangio said, this would crash the components.

Greetings
Graphefruit

ThomasBurleson commented:
@topherfangio, @EladBezalel, @DevVersion, @crisbeto, @jelbourn, @robertmesserle - please review.

DevVersion commented on docs/app/performance/internet-explorer.md:
```diff
@@ -0,0 +1,80 @@
+		})
+       .run(function($animate){
+    		$animate.enabled(!isIE);
+    	});
```
I know, I already commented about that. But it's still not correct.

This is very nitpicky, the indention here looks really weird.

```js

function isBrowserIE() {
  var ua = window.navigator.userAgent;
  return /(MSIE |Trident\/| Edge\/)/.test(ua);
}

var isIE = isBrowserIE();

angular
  .module('myApp', ['ngMaterial'])
  .config(function($mdInkRippleProvider) {
    if (isIE) {
      $mdInkRippleProvider.disableInkRipple();
    } 
  })
  .run(function($animate) {
    $animate.enabled(!isIE);
});
```

Can you please use two spaces for indention. (It should look like my code snippet)

crisbeto replied:
Didn't IE have a non-standard property along the lines of `document.mode` that you could check instead of the user agent?

david-gang replied:
I am not aware of this. I took the answer from SO http://stackoverflow.com/questions/19999388/check-if-user-is-using-ie-with-jquery

crisbeto replied:
The Angular source sniffs the IE like this `window.document.documentMode`.

david-gang replied:
ok. I will check in the angular source code

DevVersion commented on docs/app/performance/internet-explorer.md:
```diff
@@ -0,0 +1,80 @@
+angular.module('myApp', ['ngMaterial'])
+       .config(function($mdThemingProvider) {
+	           $mdThemingProvider.disableTheming();
+	});
```
The indention here is also wrong.

```js
angular
  .module('myApp', ['ngMaterial'])
  .config(function($mdThemingProvider) {
    $mdThemingProvider.disableTheming();
  });
```

crisbeto replied:
Also might be better to use ES5.

david-gang replied:
es6 is cool. why use es5 :-). This is just in the chrome tools and not run in production

crisbeto replied:
Because Material is in ES5 and this will throw errors in most browsers without transpiling.

ThomasBurleson commented:
@DevVersion - We should include the CSS fixes [shown in #8329] also.

DevVersion commented:
@ThomasBurleson Agreed. We should probably generate those style overwrites, from a script. It would be easier to maintain.

topherfangio commented on docs/app/performance/internet-explorer.md:
```diff
@@ -0,0 +1,80 @@
+
+## Tabs
+
+The [Tabs](https://material.angularjs.org/latest/demo/tabs) directive is very versatile and powerful. This comes at a certain cost of layout calculations. With Intenet Explorer this cost may be too big. An alternative to the Tabs directive could be the [Nav Bar](https://material.angularjs.org/latest/demo/navBar) directive.
```
Another thing you can note here is that you can use Tabs solely as a navigation component and switch the content out dynamically based on which tab is selected (instead of providing all of the content at once).

robertmesserle commented on docs/app/performance/internet-explorer.md:
````diff
@@ -0,0 +1,80 @@
+	.filter(e => e.attributes && e.attributes[0].nodeName==='md-theme-style')
+	.map(e => e.textContent).join("\n")
+    )
+```
````
@ThomasBurleson Could we do something to replace this step with a simple command line script?  It seems like we could do something to automatically spit this out into a CSS file.

robertmesserle replied:
Note: Not for this PR, just a note for a possibly future improvement.

crisbeto replied:
Shouldn't be too complicated to do as a Node script that can then be used with Gulp or Grunt.

david-gang commented:
Hi all,

I added a new commit with all the changes induced by your comments.
There are 2 items in todo:
*  Integrate the codepen to all codepen of material. I created two codepens to explain the difference between md-list and material. They could be integrated with all codepens of material. I don't have a pro account, so i could not upload images
 * Integrate #8329 - There were plans to integrate the changes via a script, so i just added a placeholder where the documentation comes in.

Thanks for your thorough review and comments.

BR,
David

EladBezalel commented:
lgtm

ThomasBurleson commented:
I will merge this but will not publish it on the docs yet....

Splaktar commented:
I'm adding this to the docs in 1.1.9 as part of https://github.com/angular/material/pull/11262.
